tokio/src/util/linked_list.rs

```diff
@@ -75,6 +75,15 @@ impl<T: Link> LinkedList<T> {
         }
     }
 
+    /// Creates an empty linked list
+    #[cfg(all(feature = "nightly", feature = "parking_lot", not(all(loom, test)),))]
```

Comment:
We wouldn't need the nightly feature gate if we changed the `LinkedList` type definition so that we don't need the `T: Link` bound.

Currently, it's necessary because a linked list is defined as https://github.com/tokio-rs/tokio/blob/3d53e15328c77031af8c30c6abd1eae7e772e64a/tokio/src/util/linked_list.rs#L15-L21
where the `T::Target` type is used. However, we _could_ change the definition to
```rust
pub(crate) struct LinkedList<L, T> {
    /// Linked list head
    head: Option<NonNull<T>>,

    /// Linked list tail
    tail: Option<NonNull<T>>,

    /// Node type marker
    _link: PhantomData<fn(L)>,
}
```

put the `const_new` function in an unconstrained `impl` block, like
```rust
impl<L, T> LinkedList<L, T> {
    pub(crate) const fn const_new() -> LinkedList<T> {
        LinkedList {
            head: None,
            tail: None,
        }
    }
}
```
and all the other functions (which _do_ require the `Link` bound) in a block with
```rust
impl<L: Link> LinkedList<L, L::Target> {
    // ...
}
```

This would be a much bigger change, since it would also require changing all the uses of `LinkedList`...but, it would let us have the const fn constructors on stable.

@carllerche, do you agree that not needing nightly Rust for const fn constructors is worth a somewhat larger refactoring of the linked list implementation?

Reply:
I would certainly prefer this to adding a nightly feature flag.

Reply:
I'm writing this without looking too deeply at the internals but if a refactor wouldn't be worth it couldn't we have a duplicate `LinkedList` defined like above and then use that for semaphore's internal linked list?

Reply:
> I'm writing this without looking too deeply at the internals but if a refactor wouldn't be worth it couldn't we have a duplicate `LinkedList` defined like above and then use that for semaphore's internal linked list?

I think the refactor is pretty mechanical, and IMO, maintaining two duplicate implementations of unsafe code like the linked list feels higher-risk than just making the change everywhere.